### ISSUE TITLE
Overlay singleton issues with Yarn ^1.12.0 using PnP, potentially other package managers

### ISSUE THREAD

**Issue by cdomigan** [palantir/blueprint#3193]
#### Environment

- __Package version(s)__: @blueprint/core 3.9.0 and @blueprint/select 3.3.0
- __Browser and OS versions__: Chrome 70

#### Steps to reproduce

When using Yarn PnP for package management, it is possible that multiple packages of the same version of `@blueprint/core` will exist on the filesystem when `@blueprint/select` is also installed. `@blueprint/select` then uses a different `Overlay` singleton (via `Popover`) from a distinct `@blueprint/core` installation. 

When nesting a `Select` in a `Popover`, the dismiss behaviour is then broken due to these overlays existing in separate logical stacks (ie clicking on the `Select`'s `Popover` is **always** considered an "outside click" from the point of view of the parent `Popover`.

#### Actual behavior

`Select`'s popover ends up in its own `Overlay` stack

#### Expected behavior

`Select`'s popover should use the same `Overlay` stack as the other blueprint components, and this behaviour should be guaranteed no matter which package manager is being used.

#### Possible solution

See here for discussion of this behaviour on the Yarn tracker: https://github.com/yarnpkg/yarn/issues/6534. Note that Yarn is actually performing correct behaviour here.

In general **it is not recommended to rely on module exports to implement a singleton, as they are not guaranteed to be such by the NodeJS spec**. See here: http://justjs.com/posts/singletons-in-node-js-modules-cannot-be-trusted-or-why-you-can-t-just-do-var-foo-require-baz-init, and here https://medium.com/@lazlojuly/are-node-js-modules-singletons-764ae97519af.

NPM has also recently implemented PnP-style behaviour, and I have also noticed this behaviour on some online code bins (I would need to do more testing to confirm which ones and under which circumstances).

**Comment by giladgray** [comment]
@cdomigan i'm inclined to mark this as external--you're going to have problems if you're trying to use more than one copy of blueprint at the same time. it's on you to make sure that your environment is acceptable.  this can be easily solved with a webpack `alias` to ensure that exactly one copy of a library is used at runtime.

**Comment by adidahiya** [comment]
I think this is fair criticism. The fact that blueprintjs/core relies on global state in `Overlay.openStack` in the overlay.js module is unfortunate. We should be open to a code change which at least provides an option to let the user control this global state (to solve this problem PnP users like @cdomigan). It would be less convenient than the current API, and I'm not sure how exactly it would work (we could ask the user to supply their own `openStack` array upon app initialization, but I think that runs into the same problems if there are multiple blueprintjs/core packages installed)... but open to ideas

**Comment by giladgray** [comment]
i think a better approach for the overlay stack would be to use React context so it can be implicitly passed down the tree to all nested overlays, but that's a significant feature lift.

**Comment by adidahiya** [comment]
Closing this issue as stale due to age, but noting that we still have some singleton state in Blueprint which is not ideal. Open to discussing the problem in a new issue if someone hits this again. But the general advice of "don't have multiple copies of Blueprint in your bundle" still stands.

**Comment by tnrich** [comment]
@adidahiya @giladgray 

Just ran into this "issue" myself. It was very confusing as I was switching over to use vite (from webpack v4). 

For some reason all of the popovers in my app had stopped working. I ended up running both webpack and vite side by side and putting a breakpoint here: 

<img width="1440" alt="image" src="https://github.com/palantir/blueprint/assets/2730609/affdb449-8d42-4d6d-9a66-c835e201bf06">

That's when I saw that the `overlay.openStack` between the two apps was different and landed me on the above issue. Turns out that my webpack config had already aliased blueprint like so: 

```
      "@blueprintjs/core": path.join(
        __dirname,
        "node_modules/@blueprintjs/core"
      ),
      "@blueprintjs/datetime": path.join(
        __dirname,
        "node_modules/@blueprintjs/datetime"
      ),
```

That was preventing multiple versions of blueprint from being active on the page. I think it might be helpful if some sort of warning were issued if multiple different versions of blueprint are on the page since this weird behavior can happen as a result of it.
